### DISCUSSION TITLE
I don't undertstand image relative path in Zettlr

### DISCUSSION THREAD

**Discussion by frenksisco** [Zettlr/Zettlr#4477] [Q&A]
Hi, I'm trying / testing Zettlr with my notes archive created in Obsidian, but I can't get the local images to work.

The obsidian links to my local images refer to the "attachments" subfolder of the root folder (the workspace folder in Zettlr). All the images in my notes are there, so my local image link are all like this:
`![](/attachments/image-file-name.jpg)`
or sometimes like this:
`![[image-file-name.jpg]]`
Unfortunately, as I said, none of these links show the images in Zettlr
I tried to add "/attachments" in "default image path" in Zettlr settings but it still doesn't work.
Can anyone help me to understand?

**Comment by nathanlesage** [comment]
If you start a path with a slash, that indicates an absolute path, hence it would search for a folder called "attachments" on your hard drive itself. If you mean to use the directory "attachments" relative to the file you're working in, leave out the beginning slash, or start the path with a "./" (the dot always means "start from where you are and then navigate")

Lastly, if you're still on 2.3, there is a chance that there may be bugs in the behavior of that; in that case I definitely recommend to switch to the current betas, which are extremely stable and have much more functionality.

**Reply by frenksisco** [reply]
First of all, thank you for your help.
Yes I admit I don't know much about computer language. What I would clearly like is, first, not to have to change all the links in my notes because they work in obsidian and the other markdown applications that I use; second, I want Zettlr to understand that all my images are always inside the "attachments" folder related - not to the note file I'm editing - but to the main root folder that contains all the notes an subfolders (the voult folder in Obsidian and the workspace folder in Zettlr).

I mean all my images are inside `workspace folder/attachments/`


That being said, I still haven't quite figured out what I need to write in "default image path" in Zettlr settings to make it read properly my image links

**Reply by nathanlesage** [reply]
Ah, the default image paste only applies to new images. Otherwise, the paths inside the images need to be regular paths. Obsidian here violates the standard, which means that your image paths need to be adjusted in any case (not just for Zettlr; from the top of my head I can think of VSCode, typora, Pandoc, RStudio, Ulysses, Scrivener, Bear, Notable, and likely more). I would argue it opens you many more programs if you change your paths once, and also the changes paths should then also work with Obsidian as well.

This should be quick with a find and replace command, and then all your images should work everywhere, not just Obsidian! :)

**Reply by frenksisco** [reply]
I'm sorry but I still don't understand what I have to do.
My root folder is called "note" and all my images are inside its subfolder "attachments" (note/attachments)
How should I change my links?? Because you talk about "regular path" and for me `(/attachments/image-file-name.jpg)` is a regular path

**Reply by nathanlesage** [reply]
Don't worry, paths can be confusing, and Obsidian makes it even worse, as I am going to explain. First, you are right that `/attachments/image-file-name.jpg` is indeed a regular path. I was using the word "regular" to indicate that the way Obsidian treats your images is not regular.

To understand why your images don't show in Zettlr, first some preliminaries. I'm going to provide both macOS/Linux as well as Windows examples because I'm not sure which operating system you use.

* Let us assume there is an image `image.jpg` that sits in the following folder on your computer: `/Users/frenksisco/Documents/notes/attachments/image.jpg` (macOS/Linux) or `C:\Users\frenksisco\Documents\notes\attachments\image.jpg` (Windows).
* Let us further assume that you have a note named `file.md` in which you use this image that sits in `/Users/frenksisco/Documents/notes/file.md` or `C:\Users\frenksisco\Documents\notes\file.md`

Now, you have three ways of defining the path to `image.jpg` from `file.md`.

The first one (and the "safest" one) is to use an **absolute path**. An absolute path defines the entire path to the image. In other words, you would define the image in your file like so: `![image.jpg](/Users/frenksisco/Documents/notes/attachments/image.jpg)` or `![image.jpg](C:\Users\frenksisco\Documents\notes\attachments\image.jpg)`.

The second one (and the most common one) is to use a **relative path**. A relative path defines the path to the image **from the perspective of the file**. So a relative path tells you: "If you are searching for the image, start where the file is and then follow the path". You would define a relative path to that image like so: `![image.jpg](attachments/image.jpg)` or `![image.jpg](attachments\image.jpg)`. Note that the relative path immediately starts with the folder name, and not with a slash. Another way to specify the same path would be to use `./attachments/image.jpg` or `.\attachments\image.jpg`, since the dot is the same, just more explicit.

Now to the third one. The third way is to specify the image path as you did it: `/attachments/image.jpg`. This is what Obsidian calls a "Vault location" in their docs. As you can see, this looks suspiciously like an **absolute** path, and not some relative path. To understand why that is a problem, we have to look at what happens when some app wants to **display** the image to you.

In order for any software to display that image to you, it has to load it, and to load it, the app has to find it. So it will take the path that you specified, and try to follow it to find the image.

If you specified the path **absolute**, then it is easy for the software: It literally just takes the path, checks, and sees that there is indeed an image at that place. So it can load it and show it to you.

If you specified the path **relative**, the software will notice this because your path either starts with a `.` or directly with a folder name. In that case, the software will take the folder where your note is in (i.e. `/Users/frenksisco/Documents/notes` or `C:\Users\frenksisco\Documents\notes`) and then it will add that relative path to it, which results in `/Users/frenksisco/Documents/notes/attachments/image.jpg` or `C:\Users\frenksisco\Documents\notes\attachments\image.jpg`. And we know that this is correct because there's our image.

Now here's why the third way – as Obsidian is doing it – is problematic: In Obsidian, you load a so-called "vault", which is just a fancy name for a folder. Obsidian has invented this convention where you can specify an **absolute** path, but instead of being actually absolute (as in: it is the entire path to your image), it is in fact relative – but to a very specific folder: the "vault".

So the following happens when any software encounters the following path: `/attachments/image.jpg`.

First, the software will see that the path starts with a `/`, which tells it: "Hey this is an absolute path", so it will look for a folder `/attachments` (macOS/Linux) or `C:\attachments` (Windows). However, it does not find that folder, because there is no folder named `attachments` there. Any software – including Zettlr – will now tell you "Hey, I could not find that image". This is because apps other than Obsidian do not know about this "vault location".

Obsidian has this weird thing where it will say "Okay, I could not find that image, so let us see if the path works if we just pretend that it was a relative path to the vault instead". So what Obsidian will do is first remove the `/`, then take the "vault" path, which is `/Users/frenksisco/Documents/notes` or `C:\Users\frenksisco\Documents\notes`, and add the path to it (do you see how this is basically the same as if it was a relative path?) and then sees that there is indeed an image at that location so that it can display it to you.

**The bottom line is: Any app except Obsidian does not know of this convention that an absolute path can also be treated as a relative path, and therefore no app except Obsidian will be able to find the image. Thus, you must change your image paths once, either relative, or absolute. Afterwards, your images will show up not just in Obsidian, but also in any other app.**

I hope this helps you understand what the problem you face is!

**Reply by exanrauzer** [reply]
Why wouldn't you add this to the docs?

**Comment by frenksisco** [comment]
I can't thank you enough for your patience and effort to helping me. I am actually beginning to understand what you mean. But I also better understand why Obsidian works the way it does: I think because it is a cross-platform application. I forgot to explain that for me it is an **essential factor to have a note archive that works the same way on both pc ( desktop) and android (mobile)**. Links must work in both systems and I agree with them that it is best to dedicate a specific folder for all the images and attachments, definable in the settings, with a path relative to the root folder, which after all is a very convenient starting point for searching.

I actually find extremely messy and inefficient to have image files next to the note files, or even worse, in a folder in the same location as the note.
In fact, I was hoping the "default image path" in Zettlr's settings was relative to the root folder but thanks to you now I know that it is not.

So I start to think that probably for the cross platform workflow I need is better to continue to use Obsidian or programs/apps that have a dedicated centralized attachment folder. Because in this case the image is correctly viewed on both mobile and desktop already with just the code `![[Image.jpg]]`

**Reply by nathanlesage** [reply]
I understand your sentiment here. Maybe this could be a feature that we may consider in the future!

**Reply by frenksisco** [reply]
Thanks to your explanation I am getting the related image links to work. Unfortunately, however, I just discovered something else that I just don't understand.
Zettrl perfectly recognizes image links like `![](folder/image%20file.jpg)` but it doesn't work with internal markdown links like `[Internal link](folder/note%20name.md)`. It launches an empty note named "note%20name.md" instead of the existing "note name.md." The problem seems to be the "%20" for spaces, widely used in markdown applications. For Zettlr it is okay to use it for images but not for internal links in notes... I just don't understand...

**Reply by nathanlesage** [reply]
The latter problem simply sounds like a bug that needs fixing ... which version are you currently using?

**Reply by frenksisco** [reply]
3.0.0-beta.5

**Reply by nathanlesage** [reply]
Thanks! I'll have a look!

**Reply by nathanlesage** [reply]
I have now found the issue and fixed it. It's not in beta.6 but will come with the next nightly on Monday or beta.7.

**Reply by frenksisco** [reply]
Thank you very much
